src/JSON/Tutorial.dfy

```diff
@@ -0,0 +1,269 @@
+            ""Area (km2)"": 4600
+          }, {
+            ""Name"": ""Rome"",
+            ""Founded"": -753,
+            ""Population"": 2.873e6,
+            ""Area (km2)"": 1285
+          }, {
+            ""Name"": ""Paris"",
+            ""Founded"": null,
+            ""Population"": 2.161e6,
+            ""Area (km2)"": 2383.5
+          }
+        ]
+      }");
+
+    var deserialized :- expect API.Deserialize(CITIES);
+    expect API.Serialize(deserialized) == Success(CITIES);
+
+/// Since the formatting is preserved, it is also possible to write
+/// minimally-invasive transformations over an AST.  For example, let's replace
+/// `null` in the object above with `"Unknown"`.
+///
+/// First, we construct a JSON value for the string `"Unknown"`; this could be
+/// done by hand using `View.OfBytes()`, but using `API.Deserialize` is even
+/// simpler:
+
+    var UNKNOWN :- expect API.Deserialize(Unicode.Transcode16To8(@"""Unknown"""));
+
+/// `UNKNOWN` is of type `Grammar.JSON`, which contains optional whitespace and
+/// a `Grammar.Value` under the name `UNKNOWN.t`, which we can use in the
+/// replacement:
+
+    var without_null := deserialized.(t := ReplaceNull(deserialized.t, UNKNOWN.t));
+
+/// Then, if we reserialize, we see that all formatting (and, in fact, all of
+/// the serialization work) has been reused:
+
+    expect API.Serialize(without_null) == Success(Unicode.Transcode16To8(@"{
+        ""Cities"": [
+          {
+            ""Name"": ""Boston"",
+            ""Founded"": 1630,
+            ""Population"": 689386,
+            ""Area (km2)"": 4600
+          }, {
+            ""Name"": ""Rome"",
+            ""Founded"": -753,
+            ""Population"": 2.873e6,
+            ""Area (km2)"": 1285
+          }, {
+            ""Name"": ""Paris"",
+            ""Founded"": ""Unknown"",
+            ""Population"": 2.161e6,
+            ""Area (km2)"": 2383.5
+          }
+        ]
+      }"));
+  }
+
+/// All that remains is to write the recursive traversal:
+
+  import Seq
+
+  function ReplaceNull(js: Value, replacement: Value): Value {
+    match js
+
+/// Non-recursive cases are untouched:
+
+      case Bool(_) => js
+      case String(_) => js
+      case Number(_) => js
+
+/// `Null` is replaced with the new `replacement` value:
+
+      case Null(_) => replacement
+
+/// … and objects and arrays are traversed recursively (only the data part of is
+/// traversed: other fields record information about the formatting of braces,
+/// square brackets, and whitespace, and can thus be reused without
+/// modifications):
+
+      case Object(obj) =>
+        Object(obj.(data := MapSuffixedSequence(obj.data, (s: Suffixed<jkv, jcomma>) requires s in obj.data =>
+          s.t.(v := ReplaceNull(s.t.v, replacement)))))
+      case Array(arr) =>
+        Array(arr.(data := MapSuffixedSequence(arr.data, (s: Suffixed<Value, jcomma>) requires s in arr.data =>
+          ReplaceNull(s.t, replacement))))
+  }
```

Comment:
It would not be too hard I think, but this is a specialized use case — if you're going to be writing code that touches the low-level API, you can be expected to write this yourself, I think.